File: sdk/csharp/src/Internal/Tx/Gas.cs
using Seocheon.Sdk.Constants;

namespace Seocheon.Sdk.Internal.Tx;

/// <summary>
/// Gas estimation based on message type.
/// </summary>
public static class Gas
{
    private static readonly Dictionary<string, ulong> DefaultGasLimits = new()
    {
        [Messages.TypeMsgSubmitActivity] = ChainConstants.GasSubmitActivity,
        [Messages.TypeMsgWithdrawRewards] = ChainConstants.GasWithdraw,
        [Messages.TypeMsgSend] = ChainConstants.GasSend,
    };

    /// <summary>
    /// Returns the default gas limit for a given message type.
    /// </summary>
    public static ulong GetDefaultGas(string messageTypeUrl)
    {
        return DefaultGasLimits.GetValueOrDefault(messageTypeUrl, ChainConstants.GasFallback);
    }

    /// <summary>
    /// Calculates fee from gas limit and gas price.
    /// </summary>
    public static ulong CalculateFee(ulong gasLimit, ulong gasPrice = ChainConstants.DefaultGasPrice)
    {
        return gasLimit * gasPrice;
    }

    /// <summary>
    /// Resolves gas limit: uses override if provided, otherwise uses default for message type.
    /// </summary>
    public static ulong ResolveGasLimit(string messageTypeUrl, ulong overrideGas = 0)
    {
        return overrideGas > 0 ? overrideGas : GetDefaultGas(messageTypeUrl);
    }
}

File: sdk/go/internal/tx/gas.go
package tx

// Default gas limits per message type.
const (
	DefaultGasSubmitActivity          uint64 = 200000
	DefaultGasWithdrawNodeCommission  uint64 = 300000
	DefaultGasSend                    uint64 = 100000
	DefaultGasFallback                uint64 = 200000
)

// Default fee parameters.
const (
	DefaultFeeDenom    = "uppyeo"
	DefaultGasPrice    uint64 = 250 // 250 uppyeo per gas unit
)

// DefaultGasForMessage returns the default gas limit for a given message type URL.
func DefaultGasForMessage(typeURL string) uint64 {
	switch typeURL {
	case "/seocheon.activity.v1.MsgSubmitActivity":
		return DefaultGasSubmitActivity
	case "/seocheon.node.v1.MsgWithdrawNodeCommission":
		return DefaultGasWithdrawNodeCommission
	case "/cosmos.bank.v1beta1.MsgSend":
		return DefaultGasSend
	default:
		return DefaultGasFallback
	}
}

// CalculateFee computes the fee amount from gas limit and gas price.
func CalculateFee(gasLimit, gasPrice uint64) uint64 {
	return gasLimit * gasPrice
}
